### PR TITLE
Network response mocking for Android clients

### PR DESCRIPTION
**Pull request by qichuan**
## Summary
Add network response mocking for Network plugin. See discussion [here](https://github.com/facebook/flipper/issues/475)

## Changelog
- Add Network response mocking, currently support Android clients only
- Change the Android example app to use `addInterceptor()` instead of `addNetworkInterceptor()`

## Test Plan

60549983-187ce800-9d59-11e9-8f7a-4b1b6402653d

- Connect an Android device
- Tap on Network plugin
- Click on the Mock button
- Click on Add Route button, and specify the URL 
- Edit the mock data in the text area
- Optionally, click the Headers tab to edit the headers data
- Click close button to close the dialog
- Send some network data in your application. You should be able to see the mock data appears in the Network table in those rows highlighted in yellow

### CODE REVIEW

File: android/sample/src/debug/java/com/facebook/flipper/sample/FlipperInitializer.java [hunk 1]
```diff
@@ -57,7 +57,7 @@ public static IntializationResult initFlipperPlugins(Context context, FlipperCli
 
     final OkHttpClient okHttpClient =
         new OkHttpClient.Builder()
-            .addNetworkInterceptor(interceptor)
+            .addInterceptor(interceptor)
```
Comment (henryzx) [inline android/sample/src/debug/java/com/facebook/flipper/sample/FlipperInitializer.java]: Is it possible to use network interceptor? There's many cases which app have custom interceptor, we need to mock the ultimate request.

**Comment by passy** [comment]
Hey @qichuan! Sorry for the long delay, this has somehow fallen through the cracks. This looks amazing! I'll take a look at this within the next days. :)

**Comment by passy** [comment]
@qichuan Could you perhaps rebase your stack on top of master? There are some things that have broken CI since and it would be good to get signal from Travis.

**Comment by qichuan** [comment]
> @qichuan Could you perhaps rebase your stack on top of master? There are some things that have broken CI since and it would be good to get signal from Travis.

I have done the rebase, can you please help me to re-import this PR? Thanks.

**Comment by passy** [comment]
Thank you!

**Comment by passy** [comment]
I am getting quite a few type errors when running flow on this still. Is this everything you have locally or could it be that there are some files missing from the PR?

https://gist.github.com/passy/1cf4d56eb2f252cc09292c62f87da0b0

**Comment by qichuan** [comment]
These are some lint errors, I will fix them and update the PR.

**Comment by qichuan** [comment]
Hi @passy I have fixed the lint errors. Could you please re-import the PR based on the latest commit? Thanks.

**Comment by qichuan** [comment]
Hi @passy I have fixed the lint errors, but the CI is failing due to `The command "yarn build --mac --version=$TRAVIS_BUILD_NUMBER" exited with 1.`

File: android/src/main/java/com/facebook/flipper/plugins/network/FlipperOkhttpInterceptor.java [hunk 2]
```diff
@@ -6,46 +6,89 @@
...
   @Override
   public Response intercept(Interceptor.Chain chain) throws IOException {
     Request request = chain.request();
     String identifier = UUID.randomUUID().toString();
     plugin.reportRequest(convertRequest(request, identifier));
-    Response response = chain.proceed(request);
+
+    // Check if there is a mock response
+    Response mockResponse = getMockResponse(request);
+    Response response = mockResponse != null ? getMockResponse(request) : chain.proceed(request);
```
Comment (priteshrnandgaonkar) [inline android/src/main/java/com/facebook/flipper/plugins/network/FlipperOkhttpInterceptor.java]: nit: `Response response = mockResponse != null ? mockResponse : chain.proceed(request);`

Comment (qichuan) [reply android/src/main/java/com/facebook/flipper/plugins/network/FlipperOkhttpInterceptor.java]: Sure thing, I will update the PR in a bit

**Comment by qichuan** [comment]
Hi @passy Is there anything else that I should do in order to get the PR merged?

**Comment by mrmans0n** [comment]
Wow this looks so good. Is it gonna be merged soon?

**Comment by nsekhar90** [comment]
why is this not getting merged ?

**Comment by jknoxville** [comment]
Hey @qichuan and others, I appreciate your patience, sorry that this got left behind. It was quite a task getting this into a mergeable state, as we'd converted the network plugin from flow to typescript since this PR came out, so all of these changes had to be merged with the new version of the plugin.

That said, I'd really like to get it merged, so what I've done is taken the current PR, merged it with the current typescript version of the plugin, and fixed a few small issues here and there during the conversion. You can see this updated version here: https://github.com/jknoxville/flipper/commits/network-mocking

Because the network plugin is such a core plugin, there are a few blockers to merging this in its current state, in particular:

* In place modification of props - this can cause bugs and unexpected behavior, so avoid mutating props - instead recreate objects, and use state when necessary (Examples being `routes.delete` and `checkDuplicate`).
* Let's convert the java `.*` imports to fully qualified ones, it makes code harder to read without an IDE.
* The string concatenation of URL + '|' + method is done in a few places, making it a bit error-prone. Ideally we'd make it a bit more type-safe, e.g. `Map<URL, Map<METHOD, MOCK_RESPONSE>>` and on that note, I don't think it needs to be a map, plain js objects (`{}`) would do, which will help during serialization of plugin state etc.
* I had to add a couple of `@ts-ignore`s where the types don't check. We should fix these before merging.

Some other minor issues:
* Typo in `mockResponeMap`
* Our convention is to declare all java variables as `final` when they won't be mutated.
* `@NonNull` is the default, you only need to annotate when something is Nullable.
* The parameter of `checkDuplicate` is optional, I don't think it should be.

Apologies for doing this review non-inline, I thought it would be the least intrusive way to provide the typescript converted version along with comments. For the next step, I suggest updating this PR with the network-mocking branch I linked above, and then we can go about resolving each of these issues.

**Comment by qichuan** [comment]
@jknoxville Thanks a lot for the fix. I have updated the PR with your branch and fixed the merge conflict. :) 

Hope this PR can be merged soon.

File: android/plugins/network/src/main/java/com/facebook/flipper/plugins/network/FlipperOkhttpInterceptor.java [hunk 3]
```diff
@@ -7,46 +7,83 @@
...
+import com.facebook.flipper.core.*;
+import com.facebook.flipper.plugins.common.BufferingFlipperPlugin;
 import com.facebook.flipper.plugins.network.NetworkReporter.RequestInfo;
 import com.facebook.flipper.plugins.network.NetworkReporter.ResponseInfo;
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
```
Comment (passy) [inline android/plugins/network/src/main/java/com/facebook/flipper/plugins/network/FlipperOkhttpInterceptor.java]: Can you revert these changes please? We prefer explicit imports.

Comment (passy) [inline android/plugins/network/src/main/java/com/facebook/flipper/plugins/network/FlipperOkhttpInterceptor.java]: Can you revert these changes please? We prefer explicit imports.

File: android/plugins/network/src/main/java/com/facebook/flipper/plugins/network/FlipperOkhttpInterceptor.java [hunk 4]
```diff
@@ -7,46 +7,83 @@
...
 import okio.Buffer;
 import okio.BufferedSource;
 
-public class FlipperOkhttpInterceptor implements Interceptor {
+public class FlipperOkhttpInterceptor
+    implements Interceptor, BufferingFlipperPlugin.ConnectionListener {
 
   public @Nullable NetworkFlipperPlugin plugin;
 
+  private Map<String, ResponseInfo> mockResponeMap = new HashMap<>();
```
Comment (passy) [inline android/plugins/network/src/main/java/com/facebook/flipper/plugins/network/FlipperOkhttpInterceptor.java]: *responSeMap

File: android/plugins/network/src/main/java/com/facebook/flipper/plugins/network/FlipperOkhttpInterceptor.java [hunk 5]
```diff
@@ -7,46 +7,83 @@
...
     this.plugin = null;
   }
 
   public FlipperOkhttpInterceptor(NetworkFlipperPlugin plugin) {
     this.plugin = plugin;
+    this.plugin.setConnectionListener(this);
+  }
+
+  protected void registerMockResponse(String requestUrl, String method, ResponseInfo response) {
+    mockResponeMap.put(requestUrl + "|" + method, response);
```
Comment (passy) [inline android/plugins/network/src/main/java/com/facebook/flipper/plugins/network/FlipperOkhttpInterceptor.java]: Is there a way of avoiding string concatenation here and turning this into a proper type? This seems rather fragile could break quite easily.

File: src/plugins/network/ManageMockResponsePanel.tsx [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+          ) : (
+            <TextEllipsis>{route.requestUrl}</TextEllipsis>
+          ),
+        },
+      },
+      key: index,
+    };
+  };
+
+  checkDuplicate = (routes?: Map<RequestId, Route>) => {
```
Comment (passy) [inline src/plugins/network/ManageMockResponsePanel.tsx]: The parameter here shouldn't be optional. That avoids having to have to nullable logic inside.

File: src/plugins/network/ManageMockResponsePanel.tsx [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+    const duplicateMap: {[key: string]: boolean} = {};
+    if (routes && routes.size > 0) {
+      routes.forEach((r: Route) => {
+        if (duplicateMap[r.method + '|' + r.requestUrl]) {
+          r.isDuplicate = true;
```
Comment (passy) [inline src/plugins/network/ManageMockResponsePanel.tsx]: The fundamental issue here remains that props are mutated. This is incompatible with the React model and will break. I would strongly recommend to make methods like these pure, which is both easier to reason about and will ensure you comply with the React contracts.

**Comment by qichuan** [comment]
@passy Thanks for the review! I have fixed the simple stuff (i.e., explicit import) and will work on the map and props mutation issues. 

I will let you know once everything is ready for another review.

**Comment by qichuan** [comment]
Hi @passy I have made some changes based on your suggestions, could you please take a look and let me know if there is anything else that I need to fix? Thanks a lot!

**Comment by qichuan** [comment]
Hi @passy feel free to let me know if there are any other issues that block the merge, thanks.

**Comment by qichuan** [comment]
@passy I just fixed the conflict.

I have a feeling that this PR will conflict with the base branch again if we don't' merge it soon...

**Comment by qichuan** [comment]
Hi @passy @jknoxville All I what for Christmas is having this PR merged 😂

**Comment by farukprogrammer** [comment]
please merge this PR

**Comment by nsekhar90** [comment]
Is this feature released ? Im using version 0.41.0, but do not see any 'mock' button on the network tab

**Comment by mgoyanes** [comment]
Yes it is. Using the same version (in Mac) and I can see the mock button (at the right of the clear table button) on the network tab.

**Comment by nsekhar90** [comment]
thanks for responding. I still do not see the mock button, is there anything in code we need to do to enable this feature ?

**Comment by gildor** [comment]
@mgoyanes yes, you should enable this feature in your [FlipperOkhttpInterceptor](https://github.com/facebook/flipper/blob/master/android/plugins/network/src/main/java/com/facebook/flipper/plugins/network/FlipperOkhttpInterceptor.java#L73)

File: android/sample/src/debug/java/com/facebook/flipper/sample/FlipperInitializer.java (hunk 1 above)
Comment (bangarharshit) [reply android/sample/src/debug/java/com/facebook/flipper/sample/FlipperInitializer.java:63]: Can't we just add it as the first interceptor. In that case, it will be the first interceptor.
